File: 2024/Python/src/day_04.py
from pathlib import Path
from typing import NamedTuple

type Point = tuple[int, int]
type Grid = dict[Point, str]
DIRECTIONS = ((1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (-1, -1), (-1, 1), (1, -1))


class Data(NamedTuple):
    file: list[str]
    x_pos: list[Point]
    a_pos: list[Point]
    grid: Grid


def parse() -> Data:
    file = (
        (Path(__file__).parent.parent / "inputs/day_04.txt")
        .read_text()
        .strip()
        .splitlines()
    )
    x_pos: list[Point] = []
    a_pos: list[Point] = []
    grid: Grid = {}

    for row, line in enumerate(file):
        for col, char in enumerate(line):
            grid[(col, row)] = char
            if char == "X":
                x_pos.append((col, row))
            elif char == "A":
                a_pos.append((col, row))

    return Data(file, x_pos, a_pos, grid)


def get_word(grid: Grid, start: Point, dir: Point, count: int) -> str:
    word: list[str] = []
    for i in range(count):
        grid_idx = (start[0] + dir[0] * i, start[1] + dir[1] * i)
        if grid_idx in grid:
            word.append(grid[grid_idx])

    return "".join(word)


def check_mas_x(grid: Grid, start: Point) -> bool:
    top_left = (start[0] - 1, start[1] - 1)
    word = get_word(grid, top_left, dir=(1, 1), count=3)
    if word != "MAS" and word != "SAM":
        return False

    bottom_left = (start[0] - 1, start[1] + 1)
    word = get_word(grid, bottom_left, dir=(1, -1), count=3)
    if word != "MAS" and word != "SAM":
        return False

    return True


def part_1(data: Data) -> None:
    result = 0

    for pos in data.x_pos:
        for dir in DIRECTIONS:
            if get_word(data.grid, pos, dir, len("XMAS")) == "XMAS":
                result += 1

    print(f"Day 0, Part 1: {result}")


def part_2(data: Data) -> None:
    result = sum([1 for pos in data.a_pos if check_mas_x(data.grid, pos)])
    print(f"Day 0, Part 2: {result}")


def main():
    data = parse()
    part_1(data)
    part_2(data)


if __name__ == "__main__":
    main()

File: 2024/Go/src/04/day_04.go
package main

import (
	"fmt"
	"log"
	"os"
	"strings"
	"time"
)

var DIRECTIONS []Point = []Point{
	NewPoint(1, 0),
	NewPoint(-1, 0),
	NewPoint(0, 1),
	NewPoint(0, -1),
	NewPoint(1, 1),
	NewPoint(-1, -1),
	NewPoint(-1, 1),
	NewPoint(1, -1),
}

type Point struct {
	X, Y int
}

func NewPoint(x int, y int) Point {
	return Point{X: x, Y: y}
}

type Data struct {
	Lines []string
	XPos  []Point
	APos  []Point
	Grid  map[Point]rune
}

func parse(path string) *Data {
	file, err := os.ReadFile(path)
	if err != nil {
		log.Fatalf("Error reading file %v", err)
	}
	stripped := strings.TrimSpace(string(file))
	lines := strings.Split(stripped, "\n")

	xPos := []Point{}
	aPos := []Point{}
	grid := map[Point]rune{}

	for row, line := range lines {
		for col, char := range line {
			pos := NewPoint(col, row)
			grid[pos] = char
			if char == 'X' {
				xPos = append(xPos, pos)
			} else if char == 'A' {
				aPos = append(aPos, pos)
			}
		}
	}

	return &Data{Lines: lines, XPos: xPos, APos: aPos, Grid: grid}
}

func getWord(grid map[Point]rune, start Point, dir Point, count int) string {
	word := []rune{}
	for i := range count {
		pos := NewPoint(start.X+(dir.X*i), start.Y+(dir.Y*i))
		if char, exists := grid[pos]; exists {
			word = append(word, char)
		}
	}

	return string(word)
}

func checkMasX(grid map[Point]rune, start Point) bool {
	topLeft := NewPoint(start.X-1, start.Y-1)
	dir := NewPoint(1, 1)
	word := getWord(grid, topLeft, dir, 3)
	if word != "MAS" && word != "SAM" {
		return false
	}

	bottomLeft := NewPoint(start.X-1, start.Y+1)
	dir = NewPoint(1, -1)
	word = getWord(grid, bottomLeft, dir, 3)
	if word != "MAS" && word != "SAM" {
		return false
	}

	return true
}

func part1(data *Data) {
	result := 0

	for _, pos := range data.XPos {
		for _, dir := range DIRECTIONS {
			if getWord(data.Grid, pos, dir, 4) == "XMAS" {
				result++
			}
		}
	}
	fmt.Printf("Day 04: Part 1: %v\n", result)
}

func part2(data *Data) {
	result := 0
	for _, pos := range data.APos {
		if checkMasX(data.Grid, pos) {
			result++
		}
	}
	fmt.Printf("Day 04: Part 2: %v\n", result)
}

func main() {
	fmt.Printf("------------------------------------\n")
	startTime := time.Now()
	lines := parse("inputs/day_04.txt")
	parseTime := time.Since(startTime)
	part1StartTime := time.Now()
	part1(lines)
	part1Time := time.Since(part1StartTime)
	part2StartTime := time.Now()
	part2(lines)
	part2Time := time.Since(part2StartTime)

	fmt.Printf("====================================\n")
	fmt.Printf("Finished Parsing in %v\n", parseTime)
	fmt.Printf("Finished Part 1 in %v\n", part1Time)
	fmt.Printf("Finished Part 2 in %v\n", part2Time)
	fmt.Printf("Total %v\n", time.Since(startTime))
	fmt.Printf("------------------------------------\n")
}
